designs/0026-stan-regression-syntax.md

```diff
@@ -0,0 +1,256 @@
+int G2; // number of unique members of group 1
+vector[N] y; // responses
+vector[N] x; // covariates
+int<lower=1, upper=G1> group1_idx[N]; // group 1 membership of the responses
+int<lower=1, upper=G2> group2_idx[N]; // group 2 membership of the responses
+```
+
+The parameters would look like:
+
+```
+real intercept; // population level intercept
+real slope; // population level slopes
+vector[G1] group1_intercepts; // group 1 intercepts
+vector[G1] group1_slopes; // group 1 slopes
+matrix[G1, G2] group12_intercepts; // group1:group2 intercepts
+matrix[G1, G2] group12_slopes; // group1:group2 slopes
+```
+
+The linear model can either be evaluated in a loop:
+
+```
+vector[N] linear_model;
+for(n in 1:N) {
+  linear_model[n] = intercept + slope * x[n] +
+    group1_intercepts[group1_idx[n]] + group1_slopes[group1_idx[n]] * x[n] +
+    group12_intercepts[group1_idx[n], group2_idx[n]] + group12_slopes[group1_idx[n], group2_idx[n]] * x[n];
+}
+```
+
+or it can be written in a vectorized form:
+
+```
+vector[N] linear_model = intercept + slope * x +
+    group1_intercepts[group1_idx] + group1_slopes[group1_idx] .* x +
+    diagonal(group12_intercepts[group1_idx, group2_idx]) +
+    diagonal(group12_slopes[group1_idx, group2_idx]) .* x;
+}
+```
+
+The `diagonal(...)` terms are a very inefficient way to encode the combined groupings -- in all practical cases these would be written as single group terms by actually creating a new single group index to represent the combined `group1:group2` terms. The diagonal terms suffice for demonstration though.
+
+There's really nothing to do more with the population intercept and slope. For everything else, a basic syntax that would be suitable for Stan would be:
+
+```
+covariate .* (parameters | group_idx)
+```
+
+where:
+
+1. `group_idx` is the group index of the responses. It would be an integer array of length `N` in the range `[1, G]` assuming `G` groups.
+2. `parameters` are the group level coefficients. It would be a length `G` vector-like variable.
+3. `covariate` is an optional, length `N` vector-like set of covariates.
+
+In the simplest case where `group_idx` is a 1d array of integers of length N, `parameters` is a vector of length `G`, and `covariate` is a length `N` vector of coefficients this would be equivalent to the Stan code here:
+
+```
+covariate .* parameters[group_idx]
+```
+
+This is not anything beyond what is already possible, but with a few variations on the syntax it can be.
+
+First, this syntax can support multiple-group indexing. Instead of writing loops, or doing an inefficient `covariate .* diagonal(parameters[group1, group2])`, or refactoring the two groups into one, the multiple group membership could be handled by:
+
+```
+covariate .* (parameters | group1, group2)
+```
+
+where `parameters` is now a 2d variable. To get the design matrices, the `parameters` variable could be removed.
+
+So if the following returns a length `N` vector:
+```
+covariate .* (parameters | group)
```

Comment:
Is this always going to be a `.*`? If so I think this would be easier to work with in the compiler and C++ code if we directly put the covariates in the () like 

```
(covariate, parameters | group)
```

Reply:
Yes, that seems good. I am nervous the `.*` there was actually more specialized than a `.*` in the wild might be, if that makes sense. The Fabular thing that @enetsee linked above does something like:

```
(covariate {parameters} | group)
```

Not opposed to something else either.